src/fastify-controller.js

```diff
@@ -0,0 +1,91 @@
+import { promisify } from 'util'
+import StateMachine from 'start-stop-state-machine'
+
+/**
+ * @typedef {Object} StartOpts
+ *
+ * @property {string} [host]
+ * @property {number} [port]
+ */
+
+// Class to properly manage the server lifecycle of a Fastify instance
+export class FastifyController {
+  #fastify
+  #fastifyStarted
+  #host
+  #port
+  #serverState
+
+  /**
+   * @param {Object} opts
+   * @param {import('fastify').FastifyInstance} opts.fastify
+   */
+  constructor({ fastify }) {
+    this.#fastifyStarted = false
+    this.#host = '127.0.0.1'
+    this.#port = 0
+
+    this.#fastify = fastify
+
+    this.#serverState = new StateMachine({
+      start: this.#startServer.bind(this),
+      stop: this.#stopServer.bind(this),
+    })
+  }
+
+  /**
+   * @param {StartOpts} [opts]
+   */
+  async #startServer({ host = '127.0.0.1', port = 0 } = {}) {
+    this.#host = host
+    this.#port = port
+
+    if (!this.#fastifyStarted) {
+      await this.#fastify.listen({ host: this.#host, port: this.#port })
+      this.#fastifyStarted = true
```

Comment:
I know this was copied across - I realise there is a bug here from my original code. This should avoid a race condition where calling `startServer()` twice before fastify.listen() resolves would throw an error.

```suggestion
      this.#fastifyStarted = true
      await this.#fastify.listen({ host: this.#host, port: this.#port })
```

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/440/commits/a772a425585260f9ffe436707ea95232946dece3